File: rust/quick_sort.rs
fn quick_sort(arr: &mut [i32]) {
    if arr.len() <= 1 {
        return; // base case
    }

    let pivot_index = partition(arr);

    // Recursively sort left and right partitions
    let (left, right) = arr.split_at_mut(pivot_index);
    quick_sort(left);
    quick_sort(&mut right[1..]); // skip pivot
}

fn partition(arr: &mut [i32]) -> usize {
    let len = arr.len();
    let pivot_index = len - 1; // choose last element as pivot
    let pivot = arr[pivot_index];
    let mut i = 0;

    for j in 0..pivot_index {
        if arr[j] <= pivot {
            arr.swap(i, j);
            i += 1;
        }
    }

    arr.swap(i, pivot_index);
    i
}

fn main() {
    let mut arr = [34, 7, 23, 32, 5, 62];
    println!("Before sorting: {:?}", arr);

    quick_sort(&mut arr);

    println!("After sorting:  {:?}", arr);
}

File: go/quick_sort.go
// File: quick_sort.go

package main

import "fmt"

func partition(arr []int, low, high int) int {
	pivot := arr[high]
	i := low - 1

	for j := low; j < high; j++ {
		if arr[j] <= pivot {
			i++
			arr[i], arr[j] = arr[j], arr[i]
		}
	}
	arr[i+1], arr[high] = arr[high], arr[i+1]
	return i + 1
}

func quickSort(arr []int, low, high int) {
	if low < high {
		pi := partition(arr, low, high)
		quickSort(arr, low, pi-1)
		quickSort(arr, pi+1, high)
	}
}

func main() {
	data := []int{10, 7, 8, 9, 1, 5}
	
	fmt.Println("Original:", data)
	quickSort(data, 0, len(data)-1)
	fmt.Println("Sorted:", data)
}